docs/design/proposals/codegen-refactor/centralize_1.md

```diff
@@ -0,0 +1,197 @@
+  * update accessibility of *Setters* so only the relevant interfaces are public
+    * ex: [SetResourceForStruct](https://github.com/aws-controllers-k8s/code-generator/blob/26e5da2e7656bb836ee438c05df14f2adc50197d/pkg/generate/code/set_resource.go#L1217) vs. [setResourceForContainer](https://github.com/aws-controllers-k8s/code-generator/blob/26e5da2e7656bb836ee438c05df14f2adc50197d/pkg/generate/code/set_resource.go#L1154)
+
+
+### Requirements
+* No impact to existing functionality
+* Code is easy to read, intuitive, and extendable
+* Naming in code is consistent (ex: Find vs. Get)
+* Navigating code is easier because pkg responsibility is clear
+
+
+### Prerequisites
+*code-generator* contains unit tests and also runs a subset of $SERVICE-controller e2e tests against PRs. Are these comprehensive enough to detect breakage in the code that we'll be touching/refactoring? If there are gaps between tests and code being refactored, then **tests will need to be created or updated** prior to implementing the solution.
+
+Agree on terms for consistent code. Terms such as Get vs. Find, CRDS/Resource, etc.
+
+### Implementation
+
+#### Consolidate `ackgenconfig` access and expose Getters
+The purpose is to realign responsibilities of `CRD` and `Config` making code more consistent and intuitive.
+
+1. Remove `ackgenconfig` as an attribute in `type CRD struct` in [crd.go](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/model/crd.go#L65) and fix breakage
+  * move the associated methods to `config` package and update callers, ex: [GetOutputWrapperFieldPath](https://github.com/aws-controllers-k8s/code-generator/blob/59c6892e4b61b5e2076e5e5504daba8278b82980/pkg/model/crd.go#L420):
+
+```
+// Current
+wrapperFieldPath := r.GetOutputWrapperFieldPath(op)
+
+// Proposed
+wrapperFieldPath := cfg.GetOutputWrapperFieldPath(op)
+```
+
+2. Align `Get` naming and return types in `ackgenconfig` (i.e. Get__ , not Find__)
+  * open to discussion on the exact wording/format as long as it's consistent. Proposing prepending funcs with `Get` and return the desired value and error to start
+
+```
+// Current
+func (c *Config) ResourceConfig(name string) (*ResourceConfig, bool)
+
+// Proposed
+func (c *Config) GetResourceConfig(name string) (*ResourceConfig, error)
+
+```
+
+3. Look for other wrappers and access to `ackgenconfig` and consildate to `config` package
+
+```
+// Current
+func getSortedLateInitFieldsAndConfig(cfg *ackgenconfig.Config, r *model.CRD) ([]string, map[string]*ackgenconfig.LateInitializeConfig)
+
+
+// Proposed
+func (c *Config) GetLateInitFieldsAndConfig(cfg *ackgenconfig.Config, r *model.CRD) (map[string]*ackgenconfig.LateInitializeConfig, error)
+
+// Note, this simply gets the config; sorting and any processing AFTER getting config is done by the client
```

Comment:
I disagree with the idea that we cannot have additional helper logic in the `config` package. I think it's standard practice to sanitize user input in a getter. I think the solution here is that the name should reflect the underlying data `GetLateInitFieldsAndConfig`, leaving the sorting and sanitising as an implementation detail (left in the go doc).

Reply:
> Feels like pre-optimizing the client's use case by sorting for them

I think we want as much of this stuff to run deterministically as possible, and sorting at the raw input will save us sorting in every other place where we access these values.

> Does it always make sense to sort a list of config? 

Nope. But in some places, yep. Case-by-case basis.

> Additional overhead of adding sort to every new getter (for list configs)-- both performance cost and developer effort

`O(n)` on the size of the config (probably <10) values. I would consider worrying about this early optimisation and not worth stressing.

Reply:
Aligned offline. We expect that user-configurable slices (provided in `generator.yaml`) will **never** need to *preserve order*; therefore, it should be safe/consistent/intuitive to sanitize and sort in `Getter`.

If such a config/override is needed in the future, then the order should be preserved using an index mechanism or some other abstraction more powerful than a slice.